### ISSUE TITLE
Rebase CI onto Docker Image

### ISSUE THREAD

**Issue by ryanmrichard** [NWChemEx-Project/ParallelZone#107]
This issue is more or less the acceptance test for NWChemEx-Project/.github#81.

(I'll finish filling this out once I open the issue and link it to the aforementioned issue).

**Comment by ryanmrichard** [comment]
For now just worry about the dependencies of ParallelZone. The idea is to piggy-back up the stack. So, for example, when PluginPlay pulls the ParallelZone image resulting from #109, it'll get the dependencies ParallelZone needed, plus ParallelZone. Similarly, when SimDE pulls release images of PluginPlay it'll get PluginPlay, PluginPlay's dependencies (including ParallelZone and its dependencies). And so on, up the stack.

**Comment by yzhang-23** [comment]
> For now just worry about the dependencies of ParallelZone. The idea is to piggy-back up the stack. So, for example, when PluginPlay pulls the ParallelZone image resulting from #109, it'll get the dependencies ParallelZone needed, plus ParallelZone. Similarly, when SimDE pulls release images of PluginPlay it'll get PluginPlay, PluginPlay's dependencies (including ParallelZone and its dependencies). And so on, up the stack.

So, you mean for every repo, we pull some image, either the basic image or the release image of another repo, add missing dependencies if necessary, then build/test (maybe also install?) the repo, finally we build a release image for this repo? Do I get the workflow correctly? Thanks.

**Comment by yzhang-23** [comment]
Another question I want to discuss: in the actions
-  add_licenses.yaml
-  c-cpp.yaml
-  deploy_docs.yaml
-  format.yaml
-  tag_and_release.yaml
-  test_docs.yaml
only `c-cpp.yaml`, `deploy_docs.yaml` and `test_docs.yaml` have dependencies. Other actions would not benefit from a pre-built docker image. In addition, building/testing/deploying docs only need several packages installed, so we can build a very small image for the corresponding actions. Therefore, the real task here is to build docker images which is able to facilitate the building/testing of the repos, right?  Please correct me if my understanding is wrong. Thanks.

**Comment by ryanmrichard** [comment]
> So, you mean for every repo, we pull some image, either the basic image or the release image of another repo, add missing dependencies if necessary, then build/test (maybe also install?) the repo, finally we build a release image for this repo? Do I get the workflow correctly? Thanks.

Yeah that's right. FWIW, I think we only need to install the repo when we build the release image (unless we want to test that installation works, although to some extent building the release image will do that).

> Another question I want to discuss: in the actions
> -  add_licenses.yaml
> -  c-cpp.yaml
> -  deploy_docs.yaml
> -  format.yaml
> -  tag_and_release.yaml
> -  test_docs.yaml
> only `c-cpp.yaml`, `deploy_docs.yaml` and `test_docs.yaml` have dependencies. Other actions would not benefit from a pre-built docker image. In addition, building/testing/deploying docs only need several packages installed, so we can build a very small image for the corresponding actions. Therefore, the real task here is to build docker images which is able to facilitate the building/testing of the repos, right?  Please correct me if my understanding is wrong. Thanks.

You are correct; however, for a first pass it's probably easier to just use the same base image for all workflows in a particular repo. We can go back and optimize the base image for each workflow in a subsequent pass (and only if need be; if it's negligible overhead to use the full image, it's probably not worth our time to optimize this).

**Comment by yzhang-23** [comment]
> Can't you can't just add something like:
> 
> ```miniyaml
> runs-on: ubuntu-latest
>     container:
>       image: <base_image>
> ```
> 
> to the existing workflows like `c-cpp.yaml` (possibly temporarily bypassing the re-usable workflows in the `.github` repo)?

For the building/testing task, doing it in a container is ok. I have already have an action ready for this (test pass in my forked repo but still fighting with the TOKEN issue for the NWChemEx repos); for the other tasks, such as doc publishing, currently I'm still looking for solutions inside a container.

**Comment by ryanmrichard** [comment]
What happens if you just put:
```
runs-on: ubuntu-latest
    container:
      image: <base_image>
```
at the top of `deploy_docs.yaml`?

**Comment by yzhang-23** [comment]
```
- name: Deploy to GitHub Pages
        uses: peaceiris/actions-gh-pages@v3
        with:
          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
          publish_dir: ./docs/build/html
```
This step won't run because the container has no access to other existing actions on GitHub. I think it would be too complicated if I need git clone the action repo and run it inside the container. There could be better solutions but calling me dumb I haven't find them now.

**Comment by ryanmrichard** [comment]
What's the actual error?

**Comment by yzhang-23** [comment]
> [This](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) seems to suggest it is possible, although I don't see how to do it there. Still looking.

I think the example on this page is just to launch a container and run some simple steps (no other actions on GitHub are involved).

**Comment by ryanmrichard** [comment]
> > [This](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) seems to suggest it is possible, although I don't see how to do it there. Still looking.
> 
> I think the example on this page is just to launch a container and run some simple steps (no other actions on GitHub are involved).

https://stackoverflow.com/questions/63472909/github-actions-run-steps-in-container does run a GitHub action in a container though.

**Comment by yzhang-23** [comment]
> without the error, not sure if [this](https://stackoverflow.com/questions/63472909/github-actions-run-steps-in-container) is relevant, but maybe it's an issue with how the container is made.

I could not test `deploy_docs.yaml` in the NWChemEx project for the TOKEN issue, but for other actions in my forked branch which need resources from GitHub, they stopped for no access. Let me run a direct test on `deploy_docs.yaml` in my forked branch and will post the error message here.

**Comment by yzhang-23** [comment]
> > > [This](https://docs.github.com/en/actions/using-jobs/running-jobs-in-a-container) seems to suggest it is possible, although I don't see how to do it there. Still looking.
> > 
> > 
> > I think the example on this page is just to launch a container and run some simple steps (no other actions on GitHub are involved).
> 
> https://stackoverflow.com/questions/63472909/github-actions-run-steps-in-container does run a GitHub action in a container though.

Ok, I also tested that actions/checkout@v3 works in the container created from my docker image for doc building. This is a good news! I will continue to test other actions. Thanks for pointing out this to me!

**Comment by yzhang-23** [comment]
Sorry, my previous understanding on running actions inside a container was WRONG: one CAN access GitHub actions inside a container. My previous failure of running actions inside a container have been attributed to some path issues (I'm working on fix that).

**Comment by ryanmrichard** [comment]
Glad you figured it out!

**Comment by yzhang-23** [comment]
Ok, it looks that not only me who has to fight with such path issues...please see [this link](https://github.com/actions/runner/issues/716). Finally someone suggested that one should replace ${{github.action_path}} with ${GITHUB_ACTION_PATH}, and it worked. In my eyes, this is like a BLACK MAGIC!!!

**Comment by yzhang-23** [comment]
@ryanmrichard Is it possible to create a special GitHub account and store the docker images into the registry associated with this account? I'm tackling the TOKEN issue.

**Comment by ryanmrichard** [comment]
What are you looking for, like a bot account?

**Comment by ryanmrichard** [comment]
If so, we pay for our organization's seats, so I don't think we want to create a bot account as part of the organization.

**Comment by yzhang-23** [comment]
> What are you looking for, like a bot account?

I'm currently facing such a TOKEN issue: with the provided token CONTAINER_REPO_TOKEN, I cannot push the docker image to the registry ghcr.io/${{github.actor}}. I think maybe we can store the images to a registry associated with a special account and we can generate a TOKEN to access the registry for all developers.

**Comment by yzhang-23** [comment]
> TL;DR I think we want the images stored with the NWChemEx-Project organization and `CONTAINER_REPO_TOKEN` should definitely now have sufficient permissions to do it.
> 
> Longer answer. Are you advocating for creating a bot user and storing the images in the bot user's personal registry? If so, that's probably even harder than using the NWChemEx-Project organization. First, we have to get the bot access to the private parts of NWChemEx-Project. Second, you'll have to make sure NWChemEx-Project can access the private parts of the bot user (we don't want images of private repos to be public). My guess is you'd also have to create two sets of CI (some for the bot and some for NWChemEx-Project). There's also potentially a financial issue if the private images become too large. I also think that having the images live with the NWChemEx-Project organization will instill more confidence to potential users.
> 
> Being blunt, I feel like you're probably doing something wrong. I gave `CONTAINER_REPO_TOKEN` read/write permissions to the NWChemEx-Project's container registry. Just in case those aren't the correct permissions, I have temporarily upgraded `CONTAINER_REPO_TOKEN` full admin privileges (it can do everything now). You can try again and see if that fixes it (obviously be very careful since you're more or less in sudo mode right now). If it doesn't, I don't think your problem is a permissions issue.

I'm not in favor of a bot account just in order to store the images. My problem is to find a right registry accessible to all developers in the organization --- the [GitHub documentation](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#labelling-container-images) only talks about registries associated with individual accounts ( I will appreciate greatly if someone can point out something I was missing to me). I think it is incorrect to store these images in any personal registry. So please downgrade the privilege of `CONTAINER_REPO_TOKEN`  to a safer level (including read/write to packages). Let me continue to search for other solutions. Thanks.

**Comment by ryanmrichard** [comment]
FWIW here's the organization's registry: https://github.com/orgs/NWChemEx-Project/packages. The packages in there are associated with the org, not a user. As long as `CONTAINER_REPO_TOKEN` is being used by the CI of an organization owned by NWChemEx-Project, it should have sufficient privileges to read/write packages (I downgraded it back to just read/write packages to registry)

**Comment by yzhang-23** [comment]
> FWIW here's the organization's registry: https://github.com/orgs/NWChemEx-Project/packages. The packages in there are associated with the org, not a user. As long as `CONTAINER_REPO_TOKEN` is being used by the CI of an organization owned by NWChemEx-Project, it should have sufficient privileges to read/write packages (I downgraded it back to just read/write packages to registry)

Great! Is this a public or private registry (parallelzone_gcc11 is labelled as private now)? For private registries GitHub has volume restriction, right?

**Comment by ryanmrichard** [comment]
I would have assumed that public/private is done on a per-image basis, but I don't actually know.

**Comment by yzhang-23** [comment]
> I would have assumed that public/private is done on a per-image basis, but I don't actually know.

This is difficult...one image could > 2G, which exceeds the package size restriction for private repos. I don't want to accidentally create a financial issue during my CI testing. And also I don't find a way to set the package to be public. Let me do more research. Thanks.

**Comment by ryanmrichard** [comment]
There shouldn't be any financial issues. We have disabled spending for anything other than action time. So you should get rejected if you try to exceed our storage quota.

**Comment by yzhang-23** [comment]
> There shouldn't be any financial issues. We have disabled spending for anything other than action time. So you should get rejected if you try to exceed our storage quota.

Ok, this makes me feel better. Let me run some tests and I will update. Thanks.

**Comment by ryanmrichard** [comment]
Disregard my previous statement. It looks like the spending quota for actions is also used for storage. That said, it doesn't look like we're using any of our quota even though there's three packages...

**Comment by yzhang-23** [comment]
> 

Honestly speaking, I don't quite understand GitHub policies about these quota. Anyway, I will test with caution (try not create too many large images) and minimize the chance to get charged. I will try my best!

**Comment by keceli** [comment]
As far as I know GitHub hasn't started enforcing quota limitations for private repos since they want to promote ghcr. Here is the note from this [link](https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages):

> Billing update for container image storage: The period of free use for container image storage and bandwidth for the Container registry has been extended. If you are using Container registry you'll be informed at least one month in advance of billing commencing and you'll be given an estimate of how much you should expect to pay.

**Comment by yzhang-23** [comment]
> As far as I know GitHub hasn't started enforcing quota limitations for private repos since they want to promote ghcr. Here is the note from this [link](https://docs.github.com/en/billing/managing-billing-for-github-packages/about-billing-for-github-packages):
> 
> > Billing update for container image storage: The period of free use for container image storage and bandwidth for the Container registry has been extended. If you are using Container registry you'll be informed at least one month in advance of billing commencing and you'll be given an estimate of how much you should expect to pay.

This will make me less nervous during my testing. Thanks for sharing!

**Comment by yzhang-23** [comment]
@ryanmrichard When we build the repo with the pre-built docker image, which building mode should we use, debug or release? I think we should use release, but we may run into the slow building issue and this may consume too much of our action time, right?

**Comment by ryanmrichard** [comment]
Release mode shouldn't be a problem until you need to build Mokup. I would build the images in release mode.

**Comment by yzhang-23** [comment]
@ryanmrichard Just want to confirm: the action to build the repo image (e. g., ParallelZone) should reside in the corresponding repo, not in .github, right?

**Comment by yzhang-23** [comment]
> @yzhang-23 correct, but I wouldn't worry about building an image for ParallelZone until all of the existing CI has been switched over to use the image stored in .github.

So, you mean I should be working on building the dependency images for building all the repos now, instead of actually building the release images of repos?

**Comment by ryanmrichard** [comment]
No. I meant to just focus on switching all of ParallelZone's existing CI workflows over to the base image in the .github repo. Once you have done that, then it makes sense to build a release image for ParallelZone. After that we can move on to the next repo.

**Comment by yzhang-23** [comment]
> No. I meant to just focus on switching all of ParallelZone's existing CI workflows over to the base image in the .github repo. Once you have done that, then it makes sense to build a release image for ParallelZone. After that we can move on to the next repo.

Got it. I will do that. Thanks!

**Comment by yzhang-23** [comment]
@ryanmrichard Currently I still see the building script fetching cppyy from scratch during its run. Should we put cppyy into the docker image? I think most repos need cppyy for python tests, right?

**Comment by ryanmrichard** [comment]
Yeah it's probably worth installing cppyy in the base image.

**Comment by yzhang-23** [comment]
> Yeah it's probably worth installing cppyy in the base image.

Ok, I will update the base image.

**Comment by yzhang-23** [comment]
> Yeah it's probably worth installing cppyy in the base image.
Another question: currently the base image is based on ubuntu 20.04 and python3.8(?), so if we run "pip install cppyy", the installed version may not be 3.0 (the newest). Is this ok?

**Comment by ryanmrichard** [comment]
I think we must have 3.0 of cppyy. FWIW, I think the version `pip` pulls down is by default the newest version, unless you specify otherwise.

**Comment by yzhang-23** [comment]
Just want to confirm: for the actions

- add_licenses.yaml
- format.yaml
- tag_and_release.yaml

although they have no dependencies, we still do them with the base docker image, right?

**Comment by keceli** [comment]
> I think we must have 3.0 of cppyy. FWIW, I think the version `pip` pulls down is by default the newest version, unless you specify otherwise.

I think specifying versions would be safer to avoid surprises. `pip install cppyy==3.0.0`

**Comment by yzhang-23** [comment]
> > I think we must have 3.0 of cppyy. FWIW, I think the version `pip` pulls down is by default the newest version, unless you specify otherwise.
> 
> I think specifying versions would be safer to avoid surprises. `pip install cppyy==3.0.0`

Checked in a container of the image: the version of cppyy is 3.0.0. Thanks for your suggestion!

**Comment by ryanmrichard** [comment]
![image](https://user-images.githubusercontent.com/5280139/235508101-a2855b5c-bdaa-4da8-ac8e-ac2d7c8646fd.png)

Before you do that, can you disconnect the non-image CI for this repo? As the image shows we're actually deploying the documentation twice (once with the VM-based CI and once with the image-based CI); the C++ build did something similar too. Once you do that we can close this issue.

**Comment by yzhang-23** [comment]
> Before you do that, can you disconnect the non-image CI for this repo? As the image shows we're actually deploying the documentation twice (once with the VM-based CI and once with the image-based CI); the C++ build did something similar too. Once you do that we can close this issue.

Sure. Should I simply remove the non-image CI action files, or just disable these actions?

**Comment by ryanmrichard** [comment]
@yzhang-23 I'd remove them. With version control, there's no reason to leave unwanted stuff lying around.
